### PR TITLE
WIP: Add platform common TCC library Support

### PR DESCRIPTION
**Pull request by LeanSheng**
This adds a common TCC library for the same APIs on TGL and
EHL platforms.

Signed-off-by: leanshen <lean.sheng.tan@intel.com>

### CODE REVIEW

File: Platform/ElkhartlakeBoardPkg/BoardConfig.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ def __init__(self, *args, **kwargs):
         self.FLASH_BASE_SIZE      = (self.FLASH_LAYOUT_START - self.FLASH_BASE_ADDRESS)
 
         self.HAVE_FIT_TABLE       = 1
-        self.HAVE_FSP_BIN         = 0
+        self.HAVE_FSP_BIN         = 1
```
Comment (myguitar) [inline Platform/ElkhartlakeBoardPkg/BoardConfig.py:45]: If a FSP is ready, can you do this in a separate PR?

Comment (LeanSheng) [reply Platform/ElkhartlakeBoardPkg/BoardConfig.py:45]: yeah this is a mistake. other than this, do you think the basic flow/design is fine?

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 2]
```diff
@@ -0,0 +1,639 @@
...
+
+  LdrGlobal   = (LOADER_GLOBAL_DATA *)GetLoaderGlobalDataPointer();
+  if (LdrGlobal == NULL) {
+    return EFI_NOT_FOUND;
+  }
+
+  mTccInfoHob = GetNextGuidHob (&gTccInfoHobGuid, LdrGlobal->FspHobList);
+  if (mTccInfoHob == NULL) {
+    return EFI_NOT_FOUND;
+  }
```
Comment (myguitar) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:586]: Can we do this in the platform code? Depending on platform, GUID of Hob structure may vary.

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 3]
```diff
@@ -0,0 +1,639 @@
...
+  DestAddr = (UINT8 *) PtctTable;
+  DestAddr += sizeof (EFI_ACPI_DESCRIPTION_HEADER);
+
+  for (Index = 0; Index < MaxPtctEntries; Index++) {
+    if (PtctList[Index] == NULL) {
+      continue;
+    }
+
+  CopyMem (DestAddr, PtctList[Index], PtctList[Index]->Size);
+  DestAddr += PtctList[Index]->Size;
```
Comment (myguitar) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:633]: Can we append necessary entries into PtctTable region directly without temporary memory?

File: Platform/ElkhartlakeBoardPkg/Library/TccPlatformLib/TccPlatformLib.c [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+{
+  FSPS_UPD  *FspsUpd;
+  UINT8     MaxPchPcieRootPorts;
+  UINT16    Index;
+
+  MaxPchPcieRootPorts = GetPchMaxPciePortNum ();
+  if (BiosSettings == NULL) {
+    // Set default FSP-S values for TCC mode
+    DEBUG ((DEBUG_INFO, "Setting default FSP-S TCC tuning parameters\n"));
+    FspsUpd->FspsConfig.Eist = 0; // Intel Speed Step->EnableGv
```
Comment (myguitar) [inline Platform/ElkhartlakeBoardPkg/Library/TccPlatformLib/TccPlatformLib.c:92]: FspsUpd is not initialized.

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 5]
```diff
@@ -0,0 +1,639 @@
...
+  TccBinaries->BufferBase       = TccBufferBase;
+  TccBinaries->BufferSize       = TccBufferSize;
+  TccBinaries->TuningBase       = TccTuningBase;
+  TccBinaries->TuningSize       = TccTuningSize;
+  TccBinaries->CacheconfigBase  = TccCacheconfigBase;
+  TccBinaries->CacheconfigSize  = TccCacheconfigSize;
+  TccBinaries->PtcmBase         = TccPtcmBase;
+  TccBinaries->PtcmSize         = TccPtcmSize;
+
+  Status = TccSetFspsConfig (PolicyConfig, TccBinaries);
```
Comment (myguitar) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:165]: This should be handled by platform code. FspsUpd is platform specific.

File: Platform/CommonBoardPkg/CommonBoardPkg.dec [hunk 6]
```diff
@@ -14,4 +14,5 @@
 [Includes]
   Include
 
-
+[Guids]
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/CommonBoardPkg.dec:17]: This Guid value may differ for different platforms. Please check.

File: Platform/CommonBoardPkg/Include/Library/TccLib.h [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+  union {
+    struct {
+      UINT32 Size; /* size in bytes */
+      UINT32 CacheWays; /* desired cache ways (one-hot) */
+      UINT32 CacheLevel; /* desired cache level (e.g. 2 or 3) */
+      UINT32 ValidApicIds; /* number of valid entries in u32_APIC_ID[] */
+      UINT32 ApicId[MAX_TCC_BUFFER_APIC_IDS]; /* APIC_IDs associated with this buffer */
+    } Format;
+  } Info;
+}TCC_BUFFER;
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Include/Library/TccLib.h:239]: Space between } and TCC_BUFFER

File: Platform/CommonBoardPkg/Include/Library/TccLib.h [hunk 8]
```diff
@@ -0,0 +1,526 @@
...
+
+typedef struct  {
+  UINT32 InfoFormat; /* determines which struct format to use */
+  union {
+    struct {
+      UINT32 ValidBuffers; /* number of valid entris in s_buffers[] */
+      TCC_BUFFER SBuffers[MAX_TCC_BUFFERS];
+    } Format;
+  } Info;
+}TCC_BUFFER_CONFIGURATION;
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Include/Library/TccLib.h:249]: Space here.

File: Platform/CommonBoardPkg/Include/Library/TccPlatformLib.h [hunk 9]
```diff
@@ -0,0 +1,75 @@
+/** @file
+
+  Copyright (c) 2020, Intel Corporation. All rights reserved.<BR>
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Include/Library/TccPlatformLib.h:3]: Update Year

File: Platform/CommonBoardPkg/Include/Library/TccPlatformLib.h [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+  IN BIOS_SETTINGS     *PreMemBiosSettings,
+  IN TCC_BINARIES_BASE *PreMemTccBinariesBase
+  );
+
+/**
+  This procedure will retrieve Pre-Mem TCC BIOS settings table and configure FSP-M settings
+**/
+EFI_STATUS
+TccSetFspsConfig (
+  IN BIOS_SETTINGS*     BiosSettings,
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Include/Library/TccPlatformLib.h:31]: Pointers to move to the variable name like BIOS_SETTINGS *BiosSettings.

File: Platform/CommonBoardPkg/Include/Library/TccPlatformLib.h [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+  IN TCC_BINARIES_BASE* TccBinariesBase
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Include/Library/TccPlatformLib.h:32]: Move pointer to the variable name.

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 12]
```diff
@@ -0,0 +1,639 @@
...
+#include "TccInfo.h"
+
+#define SIGNATURE_IPFW    SIGNATURE_32 ('I', 'P', 'F', 'W')
+#define SIGNATURE_TCCT    SIGNATURE_32 ('T', 'C', 'C', 'T')
+#define SIGNATURE_TCCB    SIGNATURE_32 ('T', 'C', 'C', 'B')
+#define SIGNATURE_TCCC    SIGNATURE_32 ('T', 'C', 'C', 'C')
+#define SIGNATURE_TCCM    SIGNATURE_32 ('T', 'C', 'C', 'M')
+
+TCC_INFO_HOB    *mTccInfoHob    = NULL;
+BIOS_SETTINGS   *PolicyConfig;
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:29]: If this is for this particular module, prefix with 'm'.

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 13]
```diff
@@ -0,0 +1,639 @@
...
+TCC_BINARIES_BASE *TccBinaries;
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:30]: Same comment as above.

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 14]
```diff
@@ -0,0 +1,639 @@
...
+
+  DEBUG ((DEBUG_INFO, "TccModePreMemConfig () - Start\n"));
+
+  // Load TCC tuning data from container
+  Status = LoadComponent (SIGNATURE_IPFW, SIGNATURE_TCCT, &TccPreMemTuningBase, &TccPreMemTuningSize);
+  if (EFI_ERROR (Status) || AsciiStrCmp ((CHAR8*) TccPreMemTuningBase, "dummy\0") == 0 ||
+      TccPreMemTuningSize < sizeof (TCC_STREAM_CONFIGURATION)) {
+    // Set default tuning values for TCC mode if TCC Tuning binary is not found
+    DEBUG ((DEBUG_INFO, "TCC Tuning data not found! Using default TCC tuning policies.\n"));
+    PreMemPolicyConfig               = NULL;
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:60]: Adjust spaces for assignment.

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 15]
```diff
@@ -0,0 +1,639 @@
...
+  VOID                                  *TccPreMemBufferBase;
+  UINT32                                TccPreMemBufferSize;
+  VOID                                  *TccPreMemTuningBase;
+  UINT32                                TccPreMemTuningSize;
+
+  DEBUG ((DEBUG_INFO, "TccModePreMemConfig () - Start\n"));
+
+  // Load TCC tuning data from container
+  Status = LoadComponent (SIGNATURE_IPFW, SIGNATURE_TCCT, &TccPreMemTuningBase, &TccPreMemTuningSize);
+  if (EFI_ERROR (Status) || AsciiStrCmp ((CHAR8*) TccPreMemTuningBase, "dummy\0") == 0 ||
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:56]: Is this dummy check really needed? Is it not sufficient if we check for the Stream structure size?

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 16]
```diff
@@ -0,0 +1,639 @@
...
+
+  } else {
+    // Consume TCC tuning binary settings
+    PreMemStreamConfig = (TCC_STREAM_CONFIGURATION *) TccPreMemTuningBase;
+    PreMemPolicyConfig = (BIOS_SETTINGS *) &PreMemStreamConfig->Info.Format1.BiosSettings;
+  }
+
+  // Load Tcc buffers data from container
+  Status = LoadComponent (SIGNATURE_IPFW, SIGNATURE_TCCB, &TccPreMemBufferBase, &TccPreMemBufferSize);
+  if (EFI_ERROR (Status) || AsciiStrCmp ((CHAR8*) TccPreMemBufferBase, "dummy\0") == 0) {
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:72]: If we are not checking in Tcc binaries with keyword dummy, then this dummy check may not be needed. Initially, I believe we had a dummy string originally inside the binary to differentiate the actual data. Checking for BufferBase NULL and checking for the struct size should be good enough I feel. Please verify.

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 17]
```diff
@@ -0,0 +1,639 @@
...
+**/
+
+EFI_STATUS
+EFIAPI
+TccModePostMemConfig (
+  VOID
+)
+{
+  TCC_STREAM_CONFIGURATION              *StreamConfig;
+  EFI_STATUS                            Status;
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:99]: I prefer to have this line at the end of all the declarations. Just a styling readability.

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 18]
```diff
@@ -0,0 +1,639 @@
...
+  VOID                                  *TccCacheconfigBase;
+  UINT32                                TccCacheconfigSize;
+  VOID                                  *TccPtcmBase;
+  UINT32                                TccPtcmSize;
+
+  DEBUG ((DEBUG_INFO, "TccModePostMemConfig () - Start\n"));
+
+  // Reload TCC tuning data from container
+  Status = LoadComponent (SIGNATURE_IPFW, SIGNATURE_TCCT, &TccTuningBase, &TccTuningSize);
+  if (EFI_ERROR (Status) || AsciiStrCmp ((CHAR8*) TccTuningBase, "dummy\0") == 0 ||
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:113]: Same comment as above for dummy checks.

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 19]
```diff
@@ -0,0 +1,639 @@
...
+EFI_STATUS
+EFIAPI
+GetPtctPtcdLimits (
+  IN  UINT32            PtctIndex,
+  OUT PTCT_HEADER_ENTRY **PtctEntry
+  )
+{
+  PTCT_PTCD_LIMITS                *PtctPtcdLimits;
+
+  if ((PtctEntry ==NULL) || (*PtctEntry != NULL) || (PtctIndex >= 1)) {
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:189]: Adjust space. PtctEntry == NULL

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 20]
```diff
@@ -0,0 +1,639 @@
...
+    return EFI_INVALID_PARAMETER;
+  }
+
+  PtctPtcdLimits = AllocatePool (sizeof (PTCT_PTCD_LIMITS));
+  if (PtctPtcdLimits == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  PtctPtcdLimits->Header.Size = sizeof (PTCT_PTCD_LIMITS);
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:198]: Please align the assignments from line 198 till line 210.

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 21]
```diff
@@ -0,0 +1,639 @@
...
+  DEBUG ((DEBUG_INFO, "MaxIaL2Cos: %d\n", PtctPtcdLimits->MaxIaL2Cos));
+
+  *PtctEntry = (PTCT_HEADER_ENTRY *)PtctPtcdLimits;
+  return EFI_SUCCESS;
+}
+
+/*
+  Get PTCT WRC L3 Way entry Data
+
+  @param[in]  PtctIndex   PTCT entry Index
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:221]: Adjust spacing

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 22]
```diff
@@ -0,0 +1,639 @@
...
+GetPtctWrcL3WayMasks (
+  IN  UINT32            PtctIndex,
+  OUT PTCT_HEADER_ENTRY **PtctEntry
+  )
+{
+  EFI_STATUS             Status;
+  UINT32                 WrcL3WayMasks;
+  PTCT_WRC_L3_WAY_MASKS  *PtctWrcL3WayMasks;
+
+  if ((PtctEntry ==NULL) || (*PtctEntry != NULL) || (PtctIndex >= 1)) {
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:239]: Adjust space

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 23]
```diff
@@ -0,0 +1,639 @@
...
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  PtctWrcL3WayMasks = AllocatePool (sizeof (PTCT_WRC_L3_WAY_MASKS));
+  if (PtctWrcL3WayMasks == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  PtctWrcL3WayMasks->Header.Size = sizeof (PTCT_WRC_L3_WAY_MASKS);
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:257]: Align assignments.

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 24]
```diff
@@ -0,0 +1,639 @@
...
+  DEBUG ((DEBUG_INFO, "WrcMask:0x%x\n", WrcL3WayMasks));
+
+  *PtctEntry = (PTCT_HEADER_ENTRY *)PtctWrcL3WayMasks;
+  return Status;
+}
+
+/*
+  Get PTCT GT L3 Way entry Data
+
+  @param[in]  PtctIndex         PTCT entry Index
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:270]: Spacing

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 25]
```diff
@@ -0,0 +1,639 @@
...
+  )
+{
+  EFI_STATUS            Status;
+  UINT32                GtL3WayMasks;
+  UINT8                 Index;
+  PTCT_GT_L3_WAY_MASKS  *PtctGtL3WayMasks;
+
+  Status = EFI_SUCCESS;
+
+  if ((PtctEntry ==NULL) || (*PtctEntry != NULL) || (PtctIndex >= 1)) {
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:291]: Spacing

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 26]
```diff
@@ -0,0 +1,639 @@
...
+  NumOfApic = TccBuffer->Info.Format.ValidApicIds;
+
+  // Calculate size of Buffer PTCT entry with number of Apic ids that be add to it
+  PtctEntrySize = (UINT8) (sizeof (PTCT_BUFFER) + ((NumOfApic - 1) * sizeof (UINT32)));
+  PtctBuffer = AllocatePool (PtctEntrySize);
+  if (PtctBuffer == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  PtctBuffer->Header.Size = PtctEntrySize;
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:379]: Align assignments for readability

File: Platform/CommonBoardPkg/Library/TccLib/TccLib.c [hunk 27]
```diff
@@ -0,0 +1,639 @@
...
+EFI_STATUS
+EFIAPI
+GetPtctPtcmBinary (
+  IN  UINT32            PtctIndex,
+  OUT PTCT_HEADER_ENTRY **PtctEntry
+  )
+{
+  PTCT_PTCM_BINARY                *PtctPtcmBinary;
+
+  if ((PtctEntry ==NULL) || (*PtctEntry != NULL) || (mTccInfoHob == NULL) || (PtctIndex >= 1)) {
```
Comment (sindhuragrandhi) [inline Platform/CommonBoardPkg/Library/TccLib/TccLib.c:508]: Spacing. Please review other similar lines for the rest of the file.

File: Platform/ElkhartlakeBoardPkg/Library/TccPlatformLib/TccPlatformLib.c [hunk 28]
```diff
@@ -0,0 +1,296 @@
...
+  if (PreMemBiosSettings == NULL) {
+    // Set default FSP-M values for TCC mode
+    DEBUG ((DEBUG_INFO, "Setting default FSP-M TCC tuning parameters\n"));
+    FspmUpd->FspmConfig.SaGv                  = 0;
+    FspmUpd->FspmConfig.DisPgCloseIdleTimeout = 1;
+    FspmUpd->FspmConfig.RaplLim1Ena           = 0;
+    FspmUpd->FspmConfig.RaplLim2Ena           = 0;
+    FspmUpd->FspmConfig.PowerDownMode         = 0;
+  } else {
+    //Set the FSP UPDs based on tuning data
```
Comment (sindhuragrandhi) [inline Platform/ElkhartlakeBoardPkg/Library/TccPlatformLib/TccPlatformLib.c:52]: Spacing.

File: Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/Stage2BoardInitLib.c [hunk 29]
```diff
@@ -1719,9 +1616,9 @@ UpdateFspConfig (
     }
   }
 
-  if (TCC_FEATURE_ENABLED ()) {
-    TccModePostMemConfig (FspsUpd);
-  }
+  //if (TCC_FEATURE_ENABLED ()) {
```
Comment (sindhuragrandhi) [inline Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/Stage2BoardInitLib.c:1619]: Why removing the Tcc check here? If no Tcc, we donot need to run this function at all.

File: Platform/ElkhartlakeBoardPkg/Library/TccPlatformLib/TccPlatformLib.c [hunk 30]
```diff
@@ -0,0 +1,296 @@
...
+    );
+
+  // Add 1 for max cox
+  return L2CacheRdtAlloc.Bits.HighestCosNumber + 1;
+}
+
+/**
+  Poll Run busy clear
+
+  @param[in] Base    - Base address of MMIO
```
Comment (sindhuragrandhi) [inline Platform/ElkhartlakeBoardPkg/Library/TccPlatformLib/TccPlatformLib.c:173]: Why we need "-"

File: Silicon/ElkhartlakePkg/Include/TccConfigSubRegion.h [hunk 31]
```diff
@@ -1,195 +0,0 @@
-/** @file
```
Comment (sindhuragrandhi) [inline Silicon/ElkhartlakePkg/Include/TccConfigSubRegion.h:1]: Where is this file moved to now in your patch?

**Comment by sindhuragrandhi** [comment]
Also, I'm not seeing the updated files on the TigerLake-U platform code based on the new common code. Will that be a separate patch?

**Comment by gdong1** [comment]
Close this one for now. will reopen if needed.
